#                            Default of Credit Card EDA

## 01. Importing Libraries

In [3]:
import pandas as pd
import numpy as np

## 02. Extracting Data from Excel File

In [4]:
dt = pd.read_excel('../default of credit card clients.xls', sheet_name='Data',skiprows=1)

1. look at the data 

In [5]:
dt.head(10)

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2  PAY_3  PAY_4  \
0   1      20000    2          2         1   24      2      2     -1     -1   
1   2     120000    2          2         2   26     -1      2      0      0   
2   3      90000    2          2         2   34      0      0      0      0   
3   4      50000    2          2         1   37      0      0      0      0   
4   5      50000    1          2         1   57     -1      0     -1      0   
5   6      50000    1          1         2   37      0      0      0      0   
6   7     500000    1          1         2   29      0      0      0      0   
7   8     100000    2          2         2   23      0     -1     -1      0   
8   9     140000    2          3         1   28      0      0      2      0   
9  10      20000    1          3         2   35     -2     -2     -2     -2   

   ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  PAY_AMT3  \
0  ...          0          0          0         0       689         0   
1  ...       3272       3455       3261         0      1000      1000   
2  ...      14331      14948      15549      1518      1500      1000   
3  ...      28314      28959      29547      2000      2019      1200   
4  ...      20940      19146      19131      2000     36681     10000   
5  ...      19394      19619      20024      2500      1815       657   
6  ...     542653     483003     473944     55000     40000     38000   
7  ...        221       -159        567       380       601         0   
8  ...      12211      11793       3719      3329         0       432   
9  ...          0      13007      13912         0         0         0   

   PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
0         0         0         0                           1  
1      1000         0      2000                           1  
2      1000      1000      5000                           0  
3      1100      1069      1000                           0  
4      9000       689       679                           0  
5      1000      1000       800                           0  
6     20239     13750     13770                           0  
7       581      1687      1542                           0  
8      1000      1000      1000                           0  
9     13007      1122         0                           0  

[10 rows x 25 columns]

## 03. Exploratory Data Analysis

1. What is the ratio of Male and Female in the dataset?

In [6]:
dt['SEX'].value_counts()/len(dt) *100

2    60.373333
1    39.626667
Name: SEX, dtype: float64

__OBSERVATION__:

`Females are about 60% of the overall population`

2. How many people are graduate in the dataset? 

In [6]:
dt['EDUCATION'].value_counts()

2    14030
1    10585
3     4917
5      280
4      123
6       51
0       14
Name: EDUCATION, dtype: int64

__OBSERVATION__:

`There are 10,585 graduates in the overall population`

3. What is the percentage of the married population?

In [7]:
dt['MARRIAGE'].value_counts()/len(dt) *100

2    53.213333
1    45.530000
3     1.076667
0     0.180000
Name: MARRIAGE, dtype: float64

__OBSERVATION__:

 `53% of the overall population is married.`

4. What is the minimum and maximum value of limit of balance?

In [8]:
dt['LIMIT_BAL'].max()

1000000

In [9]:
dt['LIMIT_BAL'].min()

10000

__OBSERVATION__:

`Maximum limit of balance is 1,000,000 and minimum limit of balance is 10,000` 

5. What is the minimum, maximum and average age in the dataset?

In [10]:
dt['AGE'].max()

79

In [11]:
dt['AGE'].min()

21

In [12]:
dt['AGE'].mean()

35.4855

__OBSERVATION:__

`Maximum Age is:79`
`Minimum Age is:21`
`Average Age is: 35.5`

6. Are there any missing values in the age?

In [13]:
dt['AGE'].isnull()

0        False
1        False
2        False
3        False
4        False
         ...  
29995    False
29996    False
29997    False
29998    False
29999    False
Name: AGE, Length: 30000, dtype: bool

In [14]:
dt['AGE'].isnull().sum()

0

In [15]:
dt['AGE'].notnull().sum()

30000

In [16]:
dt['AGE'].isna().sum()

0

__OBSERVATION:__

`There are no missing values in the age`

7. What is the ratio of defaulters and non-defaulters in the dataset? 

In [17]:
dt['default payment next month'].value_counts()/len(dt) *100

0    77.88
1    22.12
Name: default payment next month, dtype: float64

__OBSERVATION__:

`78%  OF THE OVERALL POPULATION ARE NON-DEFAULTERS`

8. How many levels are there in payment statuses (PAY_0 ----> PAY_5)?

dt['PAY_0'].value_counts()

In [18]:
dt['PAY_2'].value_counts()

 0    15730
-1     6050
 2     3927
-2     3782
 3      326
 4       99
 1       28
 5       25
 7       20
 6       12
 8        1
Name: PAY_2, dtype: int64

In [19]:
dt['PAY_3'].value_counts()

 0    15764
-1     5938
-2     4085
 2     3819
 3      240
 4       76
 7       27
 6       23
 5       21
 1        4
 8        3
Name: PAY_3, dtype: int64

In [20]:
dt['PAY_4'].value_counts()

 0    16455
-1     5687
-2     4348
 2     3159
 3      180
 4       69
 7       58
 5       35
 6        5
 8        2
 1        2
Name: PAY_4, dtype: int64

---------------------------

# TASK 02:

---------------------------

-------------------------------------------

9. How gender can increase the chances of  default in credit card ?

In [21]:
dt.groupby(['SEX','default payment next month'])['SEX'].count()/len(dt) *100

SEX  default payment next month
1    0                             30.050000
     1                              9.576667
2    0                             47.830000
     1                             12.543333
Name: SEX, dtype: float64

__OBSERVATION__  
`9% of the defaulters are male and 12% of the defaulters are females.`

10. How many male/female are graduated from university?

In [22]:
dt.groupby(['SEX','EDUCATION'])['SEX'].count()

SEX  EDUCATION
1    0               8
     1            4354
     2            5374
     3            1990
     4              42
     5              95
     6              25
2    0               6
     1            6231
     2            8656
     3            2927
     4              81
     5             185
     6              26
Name: SEX, dtype: int64

__OBSERVATION__  
`5374 males and 8656 females are graduated from university.`

11. How education can increase the chances of  default in credit card? 

In [23]:
dt.groupby(['default payment next month','EDUCATION'])['default payment next month'].count()

default payment next month  EDUCATION
0                           0               14
                            1             8549
                            2            10700
                            3             3680
                            4              116
                            5              262
                            6               43
1                           1             2036
                            2             3330
                            3             1237
                            4                7
                            5               18
                            6                8
Name: default payment next month, dtype: int64

__OBSERVATION__  
`Education does not have any significant affect on default payments`  
`There are many garbage values in education's data`

13. How many married couples are the defaulter? 

In [24]:
dt.groupby(['MARRIAGE','default payment next month'])['MARRIAGE'].count()

MARRIAGE  default payment next month
0         0                                49
          1                                 5
1         0                             10453
          1                              3206
2         0                             12623
          1                              3341
3         0                               239
          1                                84
Name: MARRIAGE, dtype: int64

__OBSERVATION__  
`3206 married couples are defaulters.  
0 is not defined for marital status in UCI data`

14. How many singles are graduated? 

In [25]:
dt.groupby(['MARRIAGE','EDUCATION'])['MARRIAGE'].count()

MARRIAGE  EDUCATION
0         1               4
          2               6
          3              44
1         0               4
          1            3722
          2            6842
          3            2861
          4              52
          5             150
          6              28
2         0              10
          1            6809
          2            7020
          3            1909
          4              68
          5             127
          6              21
3         1              50
          2             162
          3             103
          4               3
          5               3
          6               2
Name: MARRIAGE, dtype: int64

__OBSERVATION__  
`According to data 7020 singles are University graduates  
values 0,5,6 are not defined for education in UCI data description`

15. How limit of balance can increase/decrease the chance of default?

In [26]:
dt.groupby(['LIMIT_BAL','default payment next month'])['default payment next month'].count()

LIMIT_BAL  default payment next month
10000      0                              296
           1                              197
16000      0                                2
20000      0                             1278
           1                              698
                                         ... 
750000     0                                4
760000     0                                1
780000     0                                2
800000     0                                2
1000000    0                                1
Name: default payment next month, Length: 145, dtype: int64

16. How many married couples have delayed payment for the September 2005? 

In [27]:
dt.groupby(['MARRIAGE','PAY_5'])['MARRIAGE'].count()

MARRIAGE  PAY_5
0         -2         12
          -1         11
           0         25
           2          5
           3          1
1         -2       2279
          -1       2699
           0       7316
           2       1210
           3         82
           4         39
           5          8
           6          1
           7         24
           8          1
2         -2       2227
          -1       2778
           0       9387
           2       1389
           3         94
           4         45
           5          9
           6          3
           7         32
3         -2         28
          -1         51
           0        219
           2         22
           3          1
           7          2
Name: MARRIAGE, dtype: int64

__OBSERVATION__  
`Significan number of records are falling in the catagory of 0 and -2 which is garbage value.`

16. How age can impact the education?

In [48]:
dt["bucket"] = pd.cut(dt['AGE'], 4,labels=["Adults", "Middle-age", "Old","Extension"])

In [29]:
dt.groupby(['bucket','EDUCATION'])['AGE'].count()

bucket      EDUCATION
Adults      0               5
            1            6725
            2            8126
            3            1712
            4              78
            5             152
            6              11
Middle-age  0               8
            1            3316
            2            5113
            3            2308
            4              40
            5             109
            6              28
Old         0               1
            1             515
            2             763
            3             846
            4               5
            5              19
            6              10
Extension   1              29
            2              28
            3              51
            6               2
Name: AGE, dtype: int64

__Observation:__  


--------------------------------------------------

------------------------------------------------------

# Task 03

1. How age and martial status can collectively impact the defaulter/non-defaulter chances?

In [47]:
x = dt.groupby(['MARRIAGE','default payment next month','bucket'])['bucket'].count()
x


MARRIAGE  default payment next month  bucket    
0         0                           Adults          20
                                      Middle-age      24
                                      Old              5
          1                           Adults           2
                                      Middle-age       3
1         0                           Adults        3387
                                      Middle-age    5826
                                      Old           1172
                                      Extension       68
          1                           Adults        1033
                                      Middle-age    1719
                                      Old            429
                                      Extension       25
2         0                           Adults        9751
                                      Middle-age    2476
                                      Old            383
                                      E

__OBSERVATION__  
`Mostly Single Adults default the payments  
then comes middle aged married  
and then married adults`

2. What do you think about how education along with gender can impact the payment for month of Sept 2005 ? 

In [46]:
y= dt.groupby(['SEX','EDUCATION','PAY_0'])['PAY_0'].count()
y

SEX  EDUCATION  PAY_0
1    0          -2         2
                -1         1
                 0         3
                 1         2
     1          -2       478
                        ... 
2    5           2         8
                 3         1
     6          -1         4
                 0        19
                 2         3
Name: PAY_0, Length: 99, dtype: int64

3. Can graduates with high limit of balance default more than high school student having high balance limit? 

In [52]:
dt['Bal_Bucket']=pd.cut(dt['LIMIT_BAL'],3,labels=["Low", "Medium", "High"])

In [60]:
z= dt.groupby(['EDUCATION','default payment next month','Bal_Bucket'])['default payment next month'].count()
z

EDUCATION  default payment next month  Bal_Bucket
0          0                           Low             12
                                       Medium           2
1          0                           Low           6804
                                       Medium        1731
                                       High            14
           1                           Low           1772
                                       Medium         262
                                       High             2
2          0                           Low           9667
                                       Medium        1020
                                       High            13
           1                           Low           3171
                                       Medium         157
                                       High             2
3          0                           Low           3396
                                       Medium         280
                      

__OBSERVATION__  
`Yes  graduates with high limit of balance default more than high school student having high balance limit.`

4. Let suppose the billing amount of Sept 2005 is high do you think due to marriage (Which adds into the expenses) there will be more delay in payment Sept 2005?

In [72]:
a= dt.groupby(['PAY_0','Bal_Bucket','MARRIAGE'])['MARRIAGE'].count()
a

PAY_0  Bal_Bucket  MARRIAGE
-2     Low         0              7
                   1            999
                   2           1072
                   3             10
       Medium      1            379
                               ... 
 7     Low         2              3
       Medium      1              1
 8     Low         1              9
                   2              9
       Medium      2              1
Name: MARRIAGE, Length: 65, dtype: int64

5. How many high school students having married as their martial status opted for default?  

In [63]:
b= dt.groupby(['EDUCATION','MARRIAGE','default payment next month'])['EDUCATION'].count()
b

EDUCATION  MARRIAGE  default payment next month
0          1         0                                4
           2         0                               10
1          0         0                                4
           1         0                             2953
                     1                              769
           2         0                             5555
                     1                             1254
           3         0                               37
                     1                               13
2          0         0                                5
                     1                                1
           1         0                             5176
                     1                             1666
           2         0                             5401
                     1                             1619
           3         0                              118
                     1                               44


__OBSERVATION__  
`754`

6. What are the chances of non-default if the person whose gender is male who has graduated from some university and fall in love in with some women and then luckily got married?

In [71]:
c=dt.groupby(['default payment next month','SEX','MARRIAGE','EDUCATION'])['default payment next month'].count()
c

default payment next month  SEX  MARRIAGE  EDUCATION
0                           1    0         1               1
                                           2               1
                                           3              10
                                 1         0               2
                                           1            1300
                                                        ... 
1                           2    2         5               7
                                           6               3
                                 3         1               3
                                           2              25
                                           3              16
Name: default payment next month, Length: 77, dtype: int64

__OBSERVATION__  
`10`

7. Let say you have a new feature (column) in the data which shows whether the person has retired or still working (By the way you have to create this feature and it is called feature engineering). What are the chances of default for someone who has retired? 

In [8]:
dt['retirement']= ['retire' if x > 60 else 'working' for x in dt['AGE']] 

In [9]:
dt.head(10)

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE  PAY_0  PAY_2  PAY_3  PAY_4  \
0   1      20000    2          2         1   24      2      2     -1     -1   
1   2     120000    2          2         2   26     -1      2      0      0   
2   3      90000    2          2         2   34      0      0      0      0   
3   4      50000    2          2         1   37      0      0      0      0   
4   5      50000    1          2         1   57     -1      0     -1      0   
5   6      50000    1          1         2   37      0      0      0      0   
6   7     500000    1          1         2   29      0      0      0      0   
7   8     100000    2          2         2   23      0     -1     -1      0   
8   9     140000    2          3         1   28      0      0      2      0   
9  10      20000    1          3         2   35     -2     -2     -2     -2   

   ...  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  PAY_AMT3  PAY_AMT4  \
0  ...          0          0         0       689         0         0   
1  ...       3455       3261         0      1000      1000      1000   
2  ...      14948      15549      1518      1500      1000      1000   
3  ...      28959      29547      2000      2019      1200      1100   
4  ...      19146      19131      2000     36681     10000      9000   
5  ...      19619      20024      2500      1815       657      1000   
6  ...     483003     473944     55000     40000     38000     20239   
7  ...       -159        567       380       601         0       581   
8  ...      11793       3719      3329         0       432      1000   
9  ...      13007      13912         0         0         0     13007   

   PAY_AMT5  PAY_AMT6  default payment next month  retirement  
0         0         0                           1     working  
1         0      2000                           1     working  
2      1000      5000                           0     working  
3      1069      1000                           0     working  
4       689       679                           0     working  
5      1000       800                           0     working  
6     13750     13770                           0     working  
7      1687      1542                           0     working  
8      1000      1000                           0     working  
9      1122         0                           0     working  

[10 rows x 26 columns]

In [13]:
r=dt.groupby(['retirement','default payment next month'])['retirement'].count()/len(dt) *100
r

retirement  default payment next month
retire      0                              0.663333
            1                              0.243333
working     0                             77.216667
            1                             21.876667
Name: retirement, dtype: float64

__OBSERVATION__  
`0.24% chances of default for someone who has retired.`

8. Those who couldn't marry someone (Its so unfortunate I feel pity for them) let say they are about to retire. Most probably they won't have anyone (son or daughter or wife) so, they might think it is best opportunity for the default after all by the time police completes the investigation he/she already gone. Check it out the chances of default in such cases. 

In [19]:
dt.groupby(['SEX','retirement','MARRIAGE','default payment next month'])['default payment next month'].count()/len(dt)*100

SEX  retirement  MARRIAGE  default payment next month
1    retire      1         0                              0.290000
                           1                              0.113333
                 2         0                              0.073333
                           1                              0.013333
                 3         0                              0.006667
     working     0         0                              0.040000
                           1                              0.006667
                 1         0                             12.523333
                           1                              4.373333
                 2         0                             16.820000
                           1                              4.936667
                 3         0                              0.296667
                           1                              0.133333
2    retire      1         0                              0.260000
        

__OBSERVATION__  
`0.013% chances of default for a person a  